[Terraform] Editing index.md (#168)
elastic/terraform-provider-ec: docs/index.md

# Elastic Cloud Provider

> The Elastic Cloud provider can be used to configure and manage Elastic Cloud deployments using the Elastic Cloud

Reviewer: ```suggestion
The Elastic Cloud Terraform provider can be used to configure and manage Elastic Cloud deployments using the Elastic Cloud
```

Global comment: The full product name is `Elastic Cloud Terraform provider`, based on this [gdoc overview draft](https://docs.google.com/document/d/1hqGF2KB3mCkKJeJzvNXpQriNixr62AdOgsBEkFqxkHk/edit#) from @Kushmaro. The other short docs I put together use that same name.

## Authentication

> The Elastic Cloud (EC) provider offers two methods of authentication against the remote API: `apikey` or a combination of `username` and `password`. Depending on the environment, one or both can be used. The Public API of Elasticsearch Service (ESS) is the default `endpoint` that the provider will target.

Reviewer: ```suggestion
The Elastic Cloud Terraform provider offers two methods of authentication against the remote API: `apikey` or a combination of `username` and `password`. Depending on the environment, one or both can be used. The Public API of Elasticsearch Service (ESS) is the default `endpoint` that the provider will target.
```

I've never seen `EC` as a valid short form for Elastic Cloud and our style guide says it's _verboten_:

![image](https://user-images.githubusercontent.com/15148011/98179685-2a721c00-1eb4-11eb-82b9-83bd1ff11891.png)

> The Elastic Cloud Terraform provider offers two methods of authentication against the remote API: `apikey` or a combination of `username` and `password`. Depending on the environment, one or both can be used. The Public API of Elasticsearch Service (ESS) is the default `endpoint` that the provider will target.

Reviewer 2: I do not think that both can be used,  either `username` and `password` or `apikey` can be used, but not both.

Author: From the original text, these sentences make think that ESS accepts only apikeys, while ECE accepts both, apikeys and username+pwd:

> "Depending on the target environment, one or all can be used."
> 
> "API Keys are the recommended authentication method. They can be used when authenticating against the Elasticsearch Service (ESS) or Elastic Cloud Enterprise (ECE)"
> 
> "When targeting an ECE Installation, username and password can be used."
>

Author: If you want, we can chat/clarify on slack later today when you get online?

Reviewer 2: Ignore my comment then, I misread what you wrote, it looks good.

Elasticsearch Service (ESS) only supports `apikey`. Elastic Cloud Enterprise (ECE) supports both `apikey` and a combination of `username` and `password`.

!> **Warning:** Hard-coding credentials into any Terraform configuration is not
recommended, and risks secret leakage should this file ever be committed to a
public version control system.

> ### API Key authentication (recommended)

Reviewer: ```suggestion
### API key authentication (recommended)
```

AFAIK, `key` is not a proper noun here, based on how we refer to API keys [in our docs](https://www.elastic.co/guide/en/cloud/current/ec-api-authentication.html).

> API Keys are the recommended authentication method. They can be used to authenticate against Elasticsearch Service or Elastic Cloud Enterprise.

Reviewer: ```suggestion
API keys are the recommended authentication method. They can be used to authenticate against Elasticsearch Service or Elastic Cloud Enterprise.
```

> They can either be hardcoded in the provider `.tf` configuration (not recommended), or specified with the environment variable `EC_API_KEY`.

Reviewer: ```suggestion
They can either be specified with the `EC_API_KEY` environment variable (recommended) or hardcoded in the provider `.tf` configuration file (supported but not recommended).
```

I would put the recommended option first and mention the other option you can use at the end.

## Argument Reference

In addition to [generic `provider` arguments](https://www.terraform.io/docs/configuration/providers.html)
(for ex. `alias` and `version`), the following arguments are supported in the Elastic Cloud (EC) `provider` block:

* `endpoint` - (Optional) This is the target endpoint. It must be provided only when
   you use the Elastic Cloud provider with an ECE installation or ESS Private.

> * `apikey` - (Optional) This is the EC API Key. It is required with ESS, but it is also valid with ECE. It must be

Reviewer: ```suggestion
* `apikey` - (Optional) This is the Elastic Cloud API key. It is required with ESS, but it is also valid with ECE. It must be
```

(`EC` is not a permitted shortform and `key` is not a proper noun.)

* `insecure` - (Optional) This setting allows the provider to skip TLS verification.
  Useful when targeting installation with self-signed certificates. Not recommended when
  targeting ESS.

* `insecure` - (Optional) This setting allows the user to set a custom timeout in the

>   individual HTTP request level. Defaults to 1 minute, but might need to be tweaked if timeouts

Reviewer 3: I think we might want to specify what the default value looks like? I'll leave it up to you to decide though :)

```suggestion
  individual HTTP request level. Defaults to 1 minute (`"1m"`), but might need to be tweaked if timeouts
```

Author: Your suggestion looks much clearer. Quotes must be kept?

  are experienced.

* `verbose` - (Optional) When set to `true`, it writes a `requests.json` file in the folder
  where Terraform runs with all the outgoing HTTP requests and responses. Defaults to `false`.

* `verbose_credentials` - (Optional) When set with `verbose`, the contents of the Authorization

> header is not written. Defaults to `false`.

Reviewer 3: In this case by "will not be redacted", we meant to say that the contents of the Authorization header will be shown, that is to say the credentials will be visible, as opposed to seeing a `REDACTED` text.

```console
# This is what people will see
Authorization: ApiKey LASJKFHE5645LKJSDFASDASDASDASD==

# Instead of this
Authorization: ApiKey REDACTED
```

Author: Thanks for the clarification, Karen. If this is common terminology, then I'll revert it back to `reacted`.
